x/stakeibc/keeper/msg_server_instant_redeem_stake.go

```diff
@@ -0,0 +1,112 @@
+package keeper
+
+import (
+	"context"
+	sdkmath "cosmossdk.io/math"
+	"fmt"
+	"github.com/Stride-Labs/stride/v5/utils"
+	recordstypes "github.com/Stride-Labs/stride/v5/x/records/types"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+
+	"github.com/Stride-Labs/stride/v5/x/stakeibc/types"
+)
+
+func (k msgServer) InstantRedeemStake(goCtx context.Context, msg *types.MsgInstantRedeemStake) (*types.MsgInstantRedeemStakeResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	k.Logger(ctx).Info(fmt.Sprintf("fast unbond: %s", msg.String()))
+
+	// get our addresses, make sure they're valid
+	sender, err := sdk.AccAddressFromBech32(msg.Creator)
+	if err != nil {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "creator address is invalid: %s. err: %s", msg.Creator, err.Error())
+	}
+
+	// then make sure host zone is valid
+	hostZone, found := k.GetHostZone(ctx, msg.HostZone)
+	if !found {
+		return nil, sdkerrors.Wrapf(types.ErrInvalidHostZone, "host zone is invalid: %s", msg.HostZone)
+	}
+
+	// safety check: redemption rate must be within safety bounds
+	rateIsSafe, err := k.IsRedemptionRateWithinSafetyBounds(ctx, hostZone)
+	if !rateIsSafe || (err != nil) {
+		errMsg := fmt.Sprintf("IsRedemptionRateWithinSafetyBounds check failed. hostZone: %s, err: %s", hostZone.String(), err.Error())
+		return nil, sdkerrors.Wrapf(types.ErrRedemptionRateOutsideSafetyBounds, errMsg)
+	}
+
+	// construct desired unstaking amount from host zone
+	nativeAmount := sdk.NewDecFromInt(msg.Amount).Mul(hostZone.RedemptionRate).RoundInt()
+	// Redemption amount must not be greater than staked balance in that zone.
+	if nativeAmount.GT(hostZone.StakedBal) {
```

Comment:
Could this check ever trigger? (in cases that do not involve a hack/bug). If not, I wonder if we even need this check in `msg_server_redeem_stake`.

Might be simpler to use a stricter check below, after we sum the DRs: `nativeAmount.GT(moduleAccount balance)`.

Reply:
It shouldn't be possible other than hack/bug.  It does seem to have limited utility, I put it in here since you had it in `msg_server_redeem_stake` .

Reply:
+1 I'd bias towards replacing this with the more restrictive check `nativeAmount.GT(moduleAccount balance)`

Reply:
I thought about this some more and carefully reviewed the code.  Given that the StakedBal actually represents the amount staked in the host zone after the funds have been transferred there, and therefore no longer in the Deposit Account, this check is actually wrong and doesn't make sense for instant-redeem-stake (although usually not an issue as amount in host zone will be much larger).  Strictly speaking, and absent any bugs, the amount in the Deposit Account should always be the amount in the deposit records (which I check that it doesn't exceed immediately after) and that's it.  If there is any discrepancy, not the SendCoins call should fail if the amount to send is greater than the amount in the account.  It sounds like you are fine with simply removing this check?  Or did you want to see an explicit semi-redundant check that this amount doesn't exceed the module account balance, maybe for a better error message?

Reply:
My vote is to just remove. I think it makes sense in normal redeem stake (I believe it was in the audit report), but here it is irrelevant since only not-yet-staked funds are of concern

Reply:
Agreed with removing it from here for the reasons Kevin listed.

IMO we should also remove it from `msg_server_redeem_stake`. @sampocs why does it makes sense to keep there, even if it was listed in the audit report?

Reply:
this is now removed